src/helm/proxy/clients/huggingface_client.py

```diff
@@ -112,31 +118,42 @@ def serve_request(self, raw_request: Dict[str, Any]):
 
         return {"completions": completions, "input_length": len(encoded_input.input_ids[0])}
 
+_servers_lock: Lock = Lock()
+_servers: Dict[str, HuggingFaceServer] = {}
+
+# TODO (julian-q) server still gets instantiated multiple times
+def _get_singleton_server(model_config: HuggingFaceModelConfig) -> HuggingFaceServer:
+    with _servers_lock:
+        if model_config.model_id not in _servers:
+            hlog("WARNING: instantiating new HuggingFaceServer")
+            _servers[model_config.model_id] = HuggingFaceServer(model_config)
+    return _servers[model_config.model_id]
 
 class HuggingFaceClient(Client):
     def __init__(self, cache_config: CacheConfig):
         self.cache = Cache(cache_config)
         self.model_server_instances: Dict[str, HuggingFaceServer] = {}
 
     def get_model_server_instance(self, model) -> HuggingFaceServer:
-        if model not in self.model_server_instances:
-            model_config = get_huggingface_model_config(model)
-            if model_config:
-                self.model_server_instances[model] = HuggingFaceServer(model_config)
-            elif model == "EleutherAI/gpt-j-6B":
-                self.model_server_instances[model] = HuggingFaceServer(
-                    HuggingFaceModelConfig.from_string("EleutherAI/gpt-j-6B")
-                )
+        model_config = get_huggingface_model_config(model)
+        if not model_config:
+            organization = model.split("/")[0]
+            engine = model.split("/")[1]
+            #TODO avoid having to recreate this config in multiple places
+            if organization == "local":
+                model_config = HuggingFaceModelConfig.from_path(f"{LOCAL_MODEL_DIR}/{engine}")
+        # Special-case some models in so that users don't have to enable them with --enable-huggingface-models
+        if not model_config:
+            # Other HuggingFace hub models that we'll look up for you even if you didn't enable them via the flag
+            if model == "EleutherAI/gpt-j-6B":
```

Comment:
FYI: We eventually want to get rid of this special casing. But don't need to fix this in this PR.

Reply:
> I notice that in huggingfaceserver, all the models are loaded in "cuda:0". Since we will run the model locally (maybe with `--num-threads > 1`), maybe it's better to load the mode into different gpus?

@DavdGao Are you thinking of duplicating the model on multiple GPUs so that different threads can query different models? This sounds cool and could increase throughput. I think for now it's easier for all the threads to share one [singleton model instance](https://github.com/stanford-crfm/helm/blob/a6c2287d4c92aa542274af46aa476b9159c5fa01/src/helm/proxy/clients/huggingface_client.py#L123-L132) which would be stored on `cuda:0`. Though, we should look more into this!

If we were to load multiple model instances onto multiple GPUs here, we would want to ensure that there are enough GPUs and GPU memory to store `num_threads` instances of the model.

Another thing we can try is running multiple HELM processes in parallel, each assigned to a different GPU and a different slice of the scenario list.

Also, some very large models might require multiple GPUs to be served, in which case we might not be specifying `cuda:0` here.